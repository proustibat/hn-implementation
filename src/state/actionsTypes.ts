export enum storiesActionTypes {
  INIT = '@@INIT',
  REQUEST_STORIES = 'REQUEST_STORIES',
  REQUEST_STORIES_FAILURE = 'REQUEST_STORIES_FAILURE',
  REQUEST_STORIES_SUCCESS = 'REQUEST_STORIES_SUCCESS',
  REQUEST_STORY = 'REQUEST_STORY',
  REQUEST_STORY_FAILURE = 'REQUEST_STORY_FAILURE',
  REQUEST_STORY_SUCCESS = 'REQUEST_STORY_SUCCESS',
}
