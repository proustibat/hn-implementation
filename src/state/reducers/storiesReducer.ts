import { storiesActionTypes as Actions } from '../actionsTypes';
import { AxiosResponse } from 'axios';

export enum StoryKindType {
  job = 'job',
  story = 'story',
  comment = 'comment',
  poll = 'poll',
  pollopt = 'pollopt',
}

export type StoryProps = {
  id: number;
  deleted?: boolean; //	true if the item is deleted.
  type?: StoryKindType; // The type of item.
  by?: string; // The username of the item's author.
  time?: number; //	Creation date of the item, in Unix Time.
  text?: string; //	The comment, story or poll text. HTML.
  dead?: boolean; // true if the item is dead.
  parent?: number; // The comment's parent: either another comment or the relevant story.
  poll?: number; //	The pollopt's associated poll.
  kids?: number[]; // The ids of the item's comments, in ranked display order.
  url?: string; //	The URL of the story.
  score?: number; // The story's score, or the votes for a pollopt.
  title?: string; // The title of the story, poll or job. HTML.
  parts?: number[]; // A list of related pollopts, in display order.
  descendants?: number; //	In the case of stories or polls, the total comment count.
  loading?: boolean;
  loaded?: boolean;
};

export type StoriesReducerType = {
  items: StoryProps[];
  loading: boolean;
  loaded: boolean;
};

export const initialState: StoriesReducerType = {
  items: [],
  loading: false,
  loaded: false,
};

const LIMIT = 10;

const storiesReducer = (
  state = initialState,
  action: {
    type: Actions;
    payload?: { data?: unknown; request?: AxiosResponse };
  },
) => {
  let loadedItemId: number;
  switch (action.type) {
    case Actions.REQUEST_STORIES:
      return {
        ...state,
        loading: true,
        loaded: false,
      };
    case Actions.REQUEST_STORIES_FAILURE:
      return {
        ...state,
        loading: false,
        loaded: false,
      };
    case Actions.REQUEST_STORIES_SUCCESS:
      return {
        ...state,
        // slice temp while we don't manage queue or pagination
        items: (action?.payload?.data as number[]).slice(0, LIMIT).map(id => ({
          id,
          loaded: false,
          loading: false,
        })) as StoryProps[],
        loading: false,
        loaded: true,
      };
    case Actions.REQUEST_STORY:
      loadedItemId = action.payload?.request?.data?.id as number;
      return {
        ...state,
        items: state.items.map(item => {
          return item.id === loadedItemId
            ? {
                ...item,
                loading: true,
                loaded: false,
              }
            : item;
        }),
      };
    case Actions.REQUEST_STORY_FAILURE:
      loadedItemId = action.payload?.request?.data?.id as number;
      return {
        ...state,
        items: state.items.map(item => {
          return item.id === loadedItemId
            ? {
                ...item,
                loading: false,
                loaded: false,
              }
            : item;
        }),
      };
    case Actions.REQUEST_STORY_SUCCESS:
      const loadedItem = action.payload?.data as StoryProps;
      return {
        ...state,
        items: state.items.map(item => {
          return loadedItem && item.id === loadedItem.id
            ? {
                ...item,
                ...loadedItem,
                loading: false,
                loaded: true,
              }
            : item;
        }),
      };
    default:
      return state;
  }
};

export default storiesReducer;
