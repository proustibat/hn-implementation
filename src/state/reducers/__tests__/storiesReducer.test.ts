import storiesReducer, { initialState } from '../storiesReducer';
import { storiesActionTypes as Actions } from '../../actionsTypes';
import { stories } from '../../../fixtures/stories';
import { AxiosResponse } from 'axios';

const stateWithInitialStories = {
  items: [
    { id: 111, loaded: false, loading: false },
    { id: 222, loaded: false, loading: false },
    { id: 333, loaded: false, loading: false },
  ],
  loaded: true,
  loading: false,
};

const stateWithInitialStoriesLoadingSecond = {
  items: [
    { id: 111, loaded: false, loading: false },
    { id: 222, loaded: false, loading: true },
    { id: 333, loaded: false, loading: false },
  ],
  loaded: true,
  loading: false,
};

describe('Stories Reducer', () => {
  it('should set up default values', () => {
    // Given / When
    const state = storiesReducer(undefined, { type: Actions.INIT });

    // Then
    expect(state).toEqual(initialState);
  });

  it('should save stories and set loading and loaded booleans when requesting stories with success', () => {
    // Given
    const action = {
      type: Actions.REQUEST_STORIES_SUCCESS,
      payload: { data: stories.map(({ id }) => id) },
    };

    // When
    const state = storiesReducer(initialState, action);

    // Then
    expect(state).toMatchSnapshot();
  });

  it('should set loading state and loaded state when requesting stories begins', () => {
    // Given
    const action = { type: Actions.REQUEST_STORIES };
    const expectedStateAfterStoriesRequesting = {
      items: [],
      loading: true,
      loaded: false,
    };

    // When
    const state = storiesReducer(initialState, action);

    // Then
    expect(state).toStrictEqual(expectedStateAfterStoriesRequesting);
  });

  it('should set loading state when requesting stories fails', () => {
    // Given
    const action = { type: Actions.REQUEST_STORIES_FAILURE };
    const expectedStateAfterStoriesRequesting = {
      items: [],
      loading: false,
      loaded: false,
    };

    // When
    const state = storiesReducer(initialState, action);

    // Then
    expect(state).toStrictEqual(expectedStateAfterStoriesRequesting);
  });

  it('should save data of the requested story and set booleans correctly if success', () => {
    // Given
    const index = 1;
    const id = stateWithInitialStories.items[index].id;
    const action = {
      type: Actions.REQUEST_STORY_SUCCESS,
      payload: { data: { id } },
    };

    // When
    const state = storiesReducer(stateWithInitialStories, action);

    // Then
    expect(state.items[index]).toStrictEqual({
      id,
      loaded: true,
      loading: false,
    });
  });

  it('should set loading state of the story when requesting story begins', () => {
    // Given
    const index = 1;
    const id = stateWithInitialStories.items[index].id;
    const action = {
      type: Actions.REQUEST_STORY,
      payload: {
        request: { data: { id } } as AxiosResponse,
      },
    };

    // When
    const state = storiesReducer(stateWithInitialStories, action);

    // Then
    expect(state.items[index]).toStrictEqual({
      id,
      loaded: false,
      loading: true,
    });
  });

  it('should set loading state when requesting story fails', () => {
    // Given
    const index = 1;
    const id = stateWithInitialStories.items[index].id;
    const action = {
      type: Actions.REQUEST_STORY_FAILURE,
      payload: {
        request: { data: { id } } as AxiosResponse,
      },
    };

    // When
    const state = storiesReducer(stateWithInitialStoriesLoadingSecond, action);

    // Then
    expect(state.items[index]).toStrictEqual({
      id,
      loaded: false,
      loading: false,
    });
  });
});
