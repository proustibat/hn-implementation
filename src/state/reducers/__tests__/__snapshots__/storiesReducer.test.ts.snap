// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stories Reducer should save stories and set loading and loaded booleans when requesting stories with success 1`] = `
Object {
  "items": Array [
    Object {
      "id": 111,
      "loaded": false,
      "loading": false,
    },
    Object {
      "id": 222,
      "loaded": false,
      "loading": false,
    },
    Object {
      "id": 333,
      "loaded": false,
      "loading": false,
    },
  ],
  "loaded": true,
  "loading": false,
}
`;
